projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php

```diff
@@ -0,0 +1,518 @@
+				return __( 'Launch your blog', 'jetpack-mu-wpcom' );
+			},
+			'isLaunchTask'          => true,
+			'add_listener_callback' => 'wpcom_add_site_launch_listener',
+		),
+		'setup_blog'                      => array(
+			'get_title'            => function () {
+				return __( 'Name your blog', 'jetpack-mu-wpcom' );
+			},
+			'is_complete_callback' => 'wpcom_is_task_option_completed',
+		),
+
+		// Free plan tasks.
+		'setup_free'                      => array(
+			'get_title'            => function () {
+				return __( 'Personalize your site', 'jetpack-mu-wpcom' );
+			},
+			'is_complete_callback' => '__return_true',
+		),
+
+		// Write tasks.
+		'setup_write'                     => array(
+			'get_title'            => function () {
+				return __( 'Set up your site', 'jetpack-mu-wpcom' );
+			},
+			'is_complete_callback' => '__return_true',
+			'is_disabled_callback' => '__return_true',
+		),
+
+		// @todo Add Keep Building tasks.
+	);
+
+	$extended_task_definitions = apply_filters( 'wpcom_launchpad_extended_task_definitions', array() );
+
+	return array_merge( $extended_task_definitions, $task_definitions );
+}
+
+/**
+ * Mark a task as complete.
+ *
+ * @param string $task_id The task ID.
+ * @return bool True if the task was marked as complete, false otherwise.
+ */
+function wpcom_mark_launchpad_task_complete( $task_id ) {
+	$task_definitions = wpcom_launchpad_get_task_definitions();
+
+	// If the task ID isn't defined, return false.
+	if ( ! isset( $task_definitions[ $task_id ] ) ) {
+		return false;
+	}
+
+	// If the task has an id_map, use that instead.
+	$key = $task_id;
+	if ( isset( $task_definitions[ $task_id ]['id_map'] ) ) {
+		$key = $task_definitions[ $task_id ]['id_map'];
+	}
+
+	$statuses         = get_option( 'launchpad_checklist_tasks_statuses', array() );
+	$statuses[ $key ] = true;
+	$result           = update_option( 'launchpad_checklist_tasks_statuses', $statuses );
+
+	// Record the completion event in Tracks if we're running on WP.com.
+	if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+		require_lib( 'tracks/client' );
+
+		tracks_record_event(
+			wp_get_current_user(),
+			'launchpad_mark_task_completed',
+			array( 'task_id' => $key )
+		);
+	}
+
+	return $result;
+}
+
+/**
+ * Initialize the Launchpad task listener callbacks.
+ *
+ * @param array $task_definitions The tasks to initialize.
+ */
+function wpcom_launchpad_init_listeners( $task_definitions ) {
+	require_once WP_CONTENT_DIR . '/lib/log2logstash/log2logstash.php';
+
+	foreach ( $task_definitions as $task_id => $task_definition ) {
+		if ( isset( $task_definition['add_listener_callback'] ) && is_callable( $task_definition['add_listener_callback'] ) ) {
+			$task_data = array_merge( $task_definition, array( 'id' => $task_id ) );
+
+			try {
+				call_user_func( $task_definition['add_listener_callback'], $task_data ); // Current callbacks expect the built, registered task for the second parameter, which won't work in this case.
```

Comment:
Did you think about having the `add_listener_callback` return `true` if successful and a wp_error if it didn't succeed would be better? That way we could skip the generic `try/catch` and get a more detailed error if something went wrong.

FYI: I finished reviewing the code, I am going to run it now :)

Reply:
I hadn't considered that, and I like the idea! But I think doing so would involve refactoring all the existing callbacks to return the expected bool/error, and I see potential for disrupting those production flows in the process and drawing out our timeline. What do you think about adding it to the cleanup list as a nice-to-have once we get the Launchpad in Customer Home?

Reply:
yes, I agree with that.